http-clients/netty-nio-client/src/main/java/software/amazon/awssdk/http/nio/netty/internal/http2/MultiplexedChannelRecord.java

```diff
@@ -31,165 +27,194 @@
+            List<Http2StreamChannel> childrenToClose = new ArrayList<>(childChannels.values());
+            for (Channel childChannel : childrenToClose) {
+                childChannel.pipeline().fireExceptionCaught(t);
+            }
+        });
     }
 
-    /**
-     * Bootstraps the child stream channel and notifies the Promise on success or failure.
-     *
-     * @return BiConsumer that will bootstrap the child channel.
-     */
-    private BiConsumer<Protocol, Promise<Channel>> bootstrapChildChannel() {
-        return (s, p) -> new Http2StreamChannelBootstrap(connection)
-            .open()
-            .addListener((GenericFutureListener<Future<Http2StreamChannel>>) future -> {
-                if (future.isSuccess()) {
-                    Http2StreamChannel channel = future.getNow();
-                    childChannels.put(channel.id(), channel);
-                } else {
-                    if (!connection.isActive()) {
-                        channelReleaser.accept(connection, this);
-                    }
-                    availableStreams.incrementAndGet();
-                }
-            })
-            .addListener(promiseNotifyingListener(p));
+    public void closeAndReleaseChild(Channel childChannel) {
+        childChannel.close();
+        doInEventLoop(connection.eventLoop(), () -> {
+            childChannels.remove(childChannel.id());
+            releaseReservation();
+        });
     }
 
-    void release(Channel channel) {
-        availableStreams.incrementAndGet();
-        childChannels.remove(channel.id());
+    public Channel getConnection() {
+        return connection;
     }
 
-    boolean reusable() {
-        return !isPingInflight() && availableStreams.get() > 0 && !goAway;
-    }
+    public boolean reserveStream() {
+        for (int attempt = 0; attempt < 5; ++attempt) {
+            if (state != RecordState.OPEN || pingInFlight()) {
+                return false;
+            }
 
-    Future<Channel> getConnectionFuture() {
-        return connectionFuture;
-    }
+            long currentlyAvailable = availableChildChannels.get();
 
-    private boolean isPingInflight() {
-        // It is possible the h2 connection is not ready
-        if (connection == null) {
-            return false;
+            if (currentlyAvailable <= 0) {
+                return false;
+            }
+            if (availableChildChannels.compareAndSet(currentlyAvailable, currentlyAvailable - 1)) {
+                return true;
+            }
         }
 
+        return false;
+    }
+
+    private boolean pingInFlight() {
         return connection.attr(PING_TRACKER).get() != null;
     }
+
+    boolean canBeReleased() {
+        return state != RecordState.OPEN && availableChildChannels.get() == maxConcurrencyPerConnection;
+    }
+
+    private enum RecordState {
+        OPEN,
+        CLOSED_TO_NEW,
```

Comment:
`HALF_CLOSED`?

Reply:
HALF_CLOSED means something different in HTTP/2 world, so I'd want to avoid that phrasing. Is there something less confusing I could use?